File: 3/js/src/ShiftedBinarySearch.ts
function shiftedBinarySearch(list: number[], target: number): number {
	let start = 0;
	let end = list.length - 1;

	while (start <= end) {
		let mid = Math.floor((start + end) / 2);

		if (list[mid] == target) {
			return mid;
		}

		if (list[start] <= list[mid - 1]) {
			if (target >= list[start] && target < list[mid - 1]) {
				end = mid - 1;
			} else {
				start = mid + 1;
			}
		} else {
			if (target <= list[end] && target > list[mid + 1]) {
				start = mid + 1;
			} else {
				end = mid - 1;
			}
		}
	}

	return -1;
}

(function main() {
	const list = [45, 61, 71, 72, 73, 0, 1, 21, 33, 45];
	const target = 33;

	console.log(shiftedBinarySearch(list, target));
})();

File: 3/java/ShiftedBinarySearch.java
class ShiftedBinarySearch {
	public static int find(int arr[], int x) {
		int first = 0;
		int last = arr.length-1;
		
		while(first <= last) {
			int mid = (first+last)/2;

			if(x == arr[mid]) {
				return mid;
			} 
		
			if(arr[first] <= arr[mid]) {
				if(x >= arr[first] && x < arr[mid]) {
					last = mid-1;
				} else {
						first = mid+1;
				}
			} else if(arr[mid] <= arr[last]) {
					if(x > arr[mid] && x <= arr[last]) {
						first = mid-1;
					}	else {
							last = mid+1;
					}
			}
		}

		return -1;
	}

	public static void main(String args[]) {
		int arr[] = {4, 5, 6, 7, 0, 1, 2};	
		System.out.println(find(arr, 0));
	}
}
